### PR TITLE
Rework Tinkerbell validation logic

### PR DESCRIPTION
**Pull request by chrisdoherty4**
Update Tinkerbell validation logic reducing it to whats needed for Kubefication only. 

This redesign allows for dynamic registration of Tinkerbell Kubernetes resource validations that are run over a Tinkerbell provider managed `ClusterSpec`.

- The `internal` package is semi-temporary and serving as a staging area until we work out what pieces of the Tinkerbell provider actually need to be exported.
- Some validation was moved to the config manager related validation code but is commented out because it broke significant numbers of tests. This will be fixed in a subsequent PR. 
- All Tinkerbell service code that is no longer needed such as PBnJ clients have been removed.
- E2E tests that leveraged PBnJ have been spoofed to ensure the tests still passed. This will either require a new implementation that probes BMCs directly, or remove entirely in a subsequent PR.

### CODE REVIEW

File: pkg/providers/tinkerbell/internal/assert.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	controlPlaneMachineRef := spec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef
+	if err := validateMachineRefExists(controlPlaneMachineRef, spec.MachineConfigs); err != nil {
+		return fmt.Errorf("control plane configuration machine ref: %v", err)
+	}
+	return nil
+}
+
+// AssertEtcdMachineRefExists ensures that, if the etcd configuration is specified, it references
+// a known machine config.
+func AssertEtcdMachineRefExists(spec *ClusterSpec) error {
```
Comment (abhinavmpandey08) [inline pkg/providers/tinkerbell/internal/assert.go:46]: Should have the same validation for `templateRef` in `TinkerbellMachineConfig`

File: pkg/providers/tinkerbell/internal/validate.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+func validateMinimumExpectedHardware(cluster v1alpha1.ClusterSpec, catalogue *hardware.Catalogue) error {
+	requestedNodesCount := cluster.ControlPlaneConfiguration.Count +
+		sumWorkerNodeCounts(cluster.WorkerNodeGroupConfigurations)
+
+	// Optional external etcd configuration.
+	if cluster.ExternalEtcdConfiguration != nil {
+		requestedNodesCount += cluster.ExternalEtcdConfiguration.Count
+	}
+
+	if catalogue.TotalHardware() < requestedNodesCount {
```
Comment (abhinavmpandey08) [inline pkg/providers/tinkerbell/internal/validate.go:103]: This is fine for now but we will need to compare `requestedNodesCount` to available hardware to support the "workload cluster from existing management cluster" case

File: pkg/providers/tinkerbell/testdata/expected_results_cluster_tinkerbell_cp_stacked_etcd.yaml [hunk 3]
```diff
@@ -180,21 +180,17 @@ spec:
...
                     name: tink
+                    groups: [wheel, adm]
+                    sudo: ["ALL=(ALL) NOPASSWD:ALL"]
                     shell: /bin/bash
-                    sudo:
-                    - ALL=(ALL) NOPASSWD:ALL
+                manage_etc_hosts: localhost
                 warnings:
-                  dsid_missing_source: false
+                  dsid_missing_source: off
```
Comment (abhinavmpandey08) [inline pkg/providers/tinkerbell/testdata/expected_results_cluster_tinkerbell_cp_stacked_etcd.yaml:193]: where are these changes coming from?

**Review by abhinavmpandey08 — COMMENTED** [review]
looks good, just a jew comments

File: pkg/providers/tinkerbell/internal/validate.go (hunk 2 above)
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/internal/validate.go:103]: Yea we're not trying to address other use cases here and we may want to generalize the assertion. This just mimics the existing stuff.

File: pkg/providers/tinkerbell/testdata/expected_results_cluster_tinkerbell_cp_stacked_etcd.yaml (hunk 3 above)
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/testdata/expected_results_cluster_tinkerbell_cp_stacked_etcd.yaml:193]: They're generated. I regenerated the golden files and it found we clearly never generated them before. I assume we hand crafted them previously?

File: pkg/providers/tinkerbell/internal/assert.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	return nil
+}
+
+// AssertWorkerNodeGroupMachineRefsExists ensures all worker node group machine refs are
+// referencing a known machine config.
+func AssertWorkerNodeGroupMachineRefsExists(spec *ClusterSpec) error {
+	for _, group := range spec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		groupRef := group.MachineGroupRef
+		if err := validateMachineRefExists(groupRef, spec.MachineConfigs); err != nil {
+			return fmt.Errorf("external etcd configuration machine group ref: %v", err)
```
[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/internal/assert.go]: Fixed

File: pkg/providers/tinkerbell/internal/validate.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func validateObjectMeta(meta metav1.ObjectMeta) error {
+	if meta.Name == "" {
+		return errors.New("missing name")
+	}
+
+	if meta.Namespace == "" {
+		return fmt.Errorf("missing namespace: %v", meta.Name)
+	}
```
[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/internal/validate.go]: Dropped

**Pull request merged**